# Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Моя компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Моя первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Исследую таблицы

In [1]:
# импортирую библиотеки,установлю параметры и сохраню коннекторы
import pandas as pd
from sqlalchemy import create_engine

db_config = {'user': 'praktikum_student', 
 'pwd': 'Sdf4$2;d-d30pp', 
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, 
 'db': 'data-analyst-final-project-db'} 
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Таблица books

In [2]:
# выведу содержимое таблицы books
query = '''
            SELECT * 
            FROM books
        '''
books = pd.io.sql.read_sql(query, con = engine)
books.head(10)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

### Таблица authors

In [3]:
# выведу содержимое таблицы authors
query = '''
            SELECT * 
            FROM authors
        '''
authors = pd.io.sql.read_sql(query, con = engine)
authors.head(10)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

### Таблица publishers

In [4]:
# выведу содержимое таблицы publishers
query = '''
            SELECT * 
            FROM publishers
        '''
publishers = pd.io.sql.read_sql(query, con = engine)
publishers.head(10)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

### Таблица ratings

In [5]:
# Выведу содержимое таблицы ratings
query = '''
            SELECT * 
            FROM ratings
        '''
ratings = pd.io.sql.read_sql(query, con = engine)
ratings.head(10)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

### Таблица reviews

In [6]:
# выведу содержимое таблицы reviews
query = '''
            SELECT * 
            FROM reviews
        '''
reviews = pd.io.sql.read_sql(query, con = engine)
reviews.head(10)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

In [7]:
books.info(), authors.info(), publishers.info(), ratings.info(), reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        

(None, None, None, None, None)

<div class="alert alert-info"> <b>Коментарии студента :</b> В таблицах представлено следующие: 1) books- сожержится информация о 1000 книг, 2) authors- содержится информация о 636 авторах, 3) publishers - содержится информация о 340 издательствах, 4) ratings - содержится информация о 6456 оценок книг, 5) reviews - содержится информация о 2793 обзорах на книги

## Задания

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [8]:
pd.io.sql.read_sql('''
SELECT 
    COUNT(DISTINCT book_id) AS cnt
FROM 
    books
WHERE
    publication_date > '2000-01-01'
''', con = engine)

cnt
0  819

<div class="alert alert-info"> <b>Коментарии студента :</b>после 01.01.2000 было выпущено 819 книг

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [9]:
query = '''
SELECT
    books.title AS name, reviews_cnt AS reviews_cnt,AVG (rating) AS avg_rating
FROM 
    books LEFT JOIN (SELECT book_id, COUNT(review_id) AS reviews_cnt FROM  reviews GROUP BY  book_id) AS reviews ON reviews.book_id = books.book_id RIGHT JOIN ratings ON ratings.book_id = books.book_id GROUP BY books.title, reviews_cnt
ORDER BY avg_rating DESC
'''
pd.io.sql.read_sql(query, con = engine)

name  reviews_cnt  \
0                                  Homage to Catalonia          2.0   
1                                                March          2.0   
2    In the Hand of the Goddess (Song of the Liones...          2.0   
3                 Welcome to Temptation (Dempseys  #1)          2.0   
4    Stone of Farewell (Memory  Sorrow  and Thorn  #2)          1.0   
..                                                 ...          ...   
995  The World Is Flat: A Brief History of the Twen...          3.0   
996                                              Junky          2.0   
997                  His Excellency: George Washington          2.0   
998                                      Drowning Ruth          3.0   
999                               Harvesting the Heart          2.0   

     avg_rating  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
995        2.25  
996        2.00  
997        2.00  
998        2.00  
999        1.50  

[1000 rows x 3 columns]

<div class="alert alert-info"> <b>Коментарии студента :</b> В основном написано несколько обзоров и средний_рейтинг варьируется

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [10]:
query = '''
            SELECT b.publisher_id, p.publisher, COUNT(*) count_books
            FROM books b
            LEFT JOIN publishers p ON b.publisher_id = p.publisher_id
            WHERE num_pages > 50
            GROUP BY b.publisher_id, p.publisher
            ORDER BY COUNT(*) DESC
            LIMIT 1
        '''
pd.io.sql.read_sql(query, con = engine)

publisher_id      publisher  count_books
0           212  Penguin Books           42

<div class="alert alert-info"> <b>Коментарии студента :</b> Толще 50 страниц выпустило Пингвин Букс - 42 книги

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [11]:
query = '''
select 
       authors.author,
       authors.author_id,
       avg(ratings.rating) as avg_rating
       
FROM
    books
    INNER JOIN reviews ON reviews.book_id = books.book_id
    INNER JOIN ratings ON ratings.book_id = books.book_id
    INNER JOIN authors ON authors.author_id = books.author_id

where authors.author_id in (select 
       books.author_id
       
from books

where books.book_id in (select 
            ratings.book_id 
            
    from ratings
    group by ratings.book_id
    having count(ratings.rating_id) > 50
    ))
group by authors.author, authors.author_id     
order by avg_rating DESC    

'''
pd.io.sql.read_sql(query, con = engine) 

author  author_id  avg_rating
0                          J.K. Rowling/Mary GrandPré        236    4.293466
1                   Markus Zusak/Cao Xuân Việt Khương        402    4.264151
2                                      J.R.R. Tolkien        240    4.232892
3                                   Louisa May Alcott        376    4.196970
4                                        Rick Riordan        498    4.124481
5                                     William Golding        621    3.901408
6                                       J.D. Salinger        235    3.841727
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce        469    3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...        630    3.787879
9                                           Dan Brown        106    3.748062
10                                         Lois Lowry        372    3.741935
11            George Orwell/Boris Grabnar/Peter Škerl        195    3.729730
12                                    Stephenie Meyer        554    3.662500
13                                     John Steinbeck        311    3.628986

<div class="alert alert-info"> <b>Коментарии студента :</b> J.K. Rowling/Mary GrandPré - автор с самой высокой оценкой

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [12]:
query = '''
            SELECT AVG(count_reviews)
            FROM (
            SELECT username, COUNT(*) count_reviews
            FROM reviews
            WHERE username IN
            (
            SELECT username--, COUNT(*)
            FROM ratings
            GROUP BY username
            HAVING COUNT(*) > 50
            )
            GROUP BY username) rev
        

        '''
pd.io.sql.read_sql(query, con = engine)

avg
0  24.333333

<div class="alert alert-info"> <b>Коментарии студента :</b> среднее количество обзоров от пользователей ставивших много оценок - 24 

<div class="alert alert-info"> <b>Вывод :</b> из данного проекта можно сделать следующие заключения:
    1) В таблицах  1) books- сожержится информация о 1000 книг, 2) authors- содержится информация о 636 авторах, 3) publishers - содержится информация о 340 издательствах, 4) ratings - содержится информация о 6456 оценок книг, 5) reviews - содержится информация о 2793 обзорах на книги
    2) после 01.01.2000 было выпущено 819 книг
    3) В основном написано несколько обзоров и средний_рейтинг варьируется
    4) Толще 50 страниц выпустило Пингвин Букс - 42 книги
    5) J.K. Rowling/Mary GrandPré - автор с самой высокой оценкой
    6) среднее количество обзоров от пользователей ставивших много оценок - 24 